File: app/src/main/java/com/example/util/PhoneNumberSanitizer.kt
package com.example.util

/**
 * Restricts dialed destinations and caller IDs to E.164-ish digit strings.
 * Letters, SIP URIs, and other characters are rejected before a SIP URI is built.
 */
object PhoneNumberSanitizer {
    private val allowedChars = Regex("^[+]?\\d+$")

    /**
     * Returns a destination containing only an optional leading '+' and digits,
     * or null if the input is empty or contains anything else.
     */
    fun sanitizeDestination(raw: String): String? {
        val compact = raw.trim().filter { !it.isWhitespace() && it != '-' && it != '(' && it != ')' }
        if (compact.isEmpty()) return null
        if (!allowedChars.matches(compact)) return null
        if (compact.count { it == '+' } > 1) return null
        if (compact.contains('+') && !compact.startsWith('+')) return null
        val digits = compact.trimStart('+')
        if (digits.isEmpty()) return null
        return compact
    }

    /**
     * Filters live keypad / paste input down to an optional leading '+' and digits.
     */
    fun filterDialInput(raw: String): String {
        val builder = StringBuilder()
        raw.forEach { ch ->
            if (ch == '+' && builder.isEmpty()) {
                builder.append(ch)
            } else if (ch.isDigit()) {
                builder.append(ch)
            }
        }
        return builder.toString()
    }

    fun isValidCallerId(raw: String): Boolean {
        val sanitized = sanitizeDestination(raw) ?: return false
        return sanitized.trimStart('+').length >= 6
    }
}

File: ios/DialerID/Data/Util/PhoneNumberSanitizer.swift
import Foundation

enum PhoneNumberSanitizer {
    private static let allowed = try! NSRegularExpression(pattern: "^[+]?\\d+$")

    static func sanitizeDestination(_ raw: String) -> String? {
        let compact = raw.trimmingCharacters(in: .whitespacesAndNewlines)
            .filter { !$0.isWhitespace && $0 != "-" && $0 != "(" && $0 != ")" }
        if compact.isEmpty { return nil }
        let range = NSRange(compact.startIndex..., in: compact)
        guard allowed.firstMatch(in: compact, range: range) != nil else { return nil }
        if compact.filter({ $0 == "+" }).count > 1 { return nil }
        if compact.contains("+") && !compact.hasPrefix("+") { return nil }
        let digits = compact.trimmingCharacters(in: CharacterSet(charactersIn: "+"))
        if digits.isEmpty { return nil }
        return compact
    }

    static func filterDialInput(_ raw: String) -> String {
        var builder = ""
        for ch in raw {
            if ch == "+" && builder.isEmpty {
                builder.append(ch)
            } else if ch.isNumber {
                builder.append(ch)
            }
        }
        return builder
    }

    static func isValidCallerId(_ raw: String) -> Bool {
        guard let sanitized = sanitizeDestination(raw) else { return false }
        return sanitized.trimmingCharacters(in: CharacterSet(charactersIn: "+")).count >= 6
    }
}
